runtime/caml/eventring.h

```diff
@@ -0,0 +1,300 @@
+    EV_EXPLICIT_GC_MINOR,
+    EV_EXPLICIT_GC_MAJOR,
+    EV_EXPLICIT_GC_FULL_MAJOR,
+    EV_EXPLICIT_GC_COMPACT,
+    EV_MAJOR,
+    EV_MAJOR_ROOTS,
+    EV_MAJOR_SWEEP,
+    EV_MAJOR_MARK_ROOTS,
+    EV_MAJOR_MARK_MAIN,
+    EV_MAJOR_MARK_FINAL,
+    EV_MAJOR_MARK,
+    EV_MAJOR_MARK_GLOBAL_ROOTS_SLICE,
+    EV_MAJOR_ROOTS_GLOBAL,
+    EV_MAJOR_ROOTS_DYNAMIC_GLOBAL,
+    EV_MAJOR_ROOTS_LOCAL,
+    EV_MAJOR_ROOTS_C,
+    EV_MAJOR_ROOTS_FINALISED,
+    EV_MAJOR_ROOTS_MEMPROF,
+    EV_MAJOR_ROOTS_HOOK,
+    EV_MAJOR_CHECK_AND_COMPACT,
+    EV_MINOR,
+    EV_MINOR_LOCAL_ROOTS,
+    EV_MINOR_REF_TABLES,
+    EV_MINOR_COPY,
+    EV_MINOR_UPDATE_WEAK,
+    EV_MINOR_FINALIZED,
+    EV_EXPLICIT_GC_MAJOR_SLICE,
+    EV_DOMAIN_SEND_INTERRUPT,
+    EV_DOMAIN_IDLE_WAIT,
+    EV_FINALISE_UPDATE_FIRST,
+    EV_FINALISE_UPDATE_LAST,
+    EV_INTERRUPT_REMOTE,
+    EV_MAJOR_EPHE_MARK,
+    EV_MAJOR_EPHE_SWEEP,
+    EV_MAJOR_FINISH_MARKING,
+    EV_MAJOR_GC_CYCLE_DOMAINS,
+    EV_MAJOR_GC_PHASE_CHANGE,
+    EV_MAJOR_GC_STW,
+    EV_MAJOR_MARK_OPPORTUNISTIC,
+    EV_MAJOR_SLICE,
+    EV_MINOR_CLEAR,
+    EV_MINOR_FINALIZERS_OLDIFY,
+    EV_MINOR_GLOBAL_ROOTS,
+    EV_MINOR_LEAVE_BARRIER,
+    EV_STW_API_BARRIER,
+    EV_STW_HANDLER,
+    EV_STW_LEADER,
+    EV_MAJOR_FINISH_SWEEPING,
+    EV_MINOR_FINALIZERS_ADMIN,
+    EV_MINOR_REMEMBERED_SET,
+    EV_MINOR_REMEMBERED_SET_PROMOTE,
+    EV_MINOR_LOCAL_ROOTS_PROMOTE,
+    EV_DOMAIN_CONDITION_WAIT
+} ev_runtime_phase;
+
+typedef enum {
+    EV_C_ALLOC_JUMP,
+    EV_C_FORCE_MINOR_ALLOC_SMALL,
+    EV_C_FORCE_MINOR_MAKE_VECT,
+    EV_C_FORCE_MINOR_SET_MINOR_HEAP_SIZE,
+    EV_C_FORCE_MINOR_WEAK,
+    EV_C_FORCE_MINOR_MEMPROF,
+    EV_C_MAJOR_MARK_SLICE_REMAIN,
+    EV_C_MAJOR_MARK_SLICE_FIELDS,
+    EV_C_MAJOR_MARK_SLICE_POINTERS,
+    EV_C_MAJOR_WORK_EXTRA,
+    EV_C_MAJOR_WORK_MARK,
+    EV_C_MAJOR_WORK_SWEEP,
+    EV_C_MINOR_PROMOTED,
+    EV_C_MINOR_ALLOCATED,
+    EV_C_REQUEST_MAJOR_ALLOC_SHR,
+    EV_C_REQUEST_MAJOR_ADJUST_GC_SPEED,
+    EV_C_REQUEST_MINOR_REALLOC_REF_TABLE,
+    EV_C_REQUEST_MINOR_REALLOC_EPHE_REF_TABLE,
+    EV_C_REQUEST_MINOR_REALLOC_CUSTOM_TABLE
+} ev_runtime_counter;
+
+/* external C-API for reading from the eventring */
+struct caml_eventring_cursor;
+
+typedef enum {
+  E_SUCCESS = 0,
+  E_CURSOR_NOT_OPEN = -1,
+  E_CORRUPT_STREAM = -2,
+  E_ALLOC_FAIL = -3,
+  E_PATH_FAILURE = -4,
+  E_OPEN_FAILURE = -5,
+  E_NO_CURRENT_RING = -6,
+  E_MAP_FAILURE = -7,
+} eventring_error;
+
+/* Starts eventring. Needs to be called before [caml_eventring_create_cursor] */
+extern value caml_eventring_start();
+
+/* Pauses eventring. No new events (other than the pause itself) will be written
+   to the eventrings by this domain immediately and all other domains soon. */
+extern value caml_eventring_pause();
+
+/* Removes eventring. New events (as well as a resume event) will be written to
+   this domain immediately and all other domains soon. */
+extern value caml_eventring_resume();
```

Comment:
Yes, that's correct in that it's up to the OS in eventring's case. If you put the eventrings on the filesystem it will generate writes, though it's possible to place them in /dev/shm (atleast on Linux) and have them remain fully in-memory.

I mostly included pause/resume because Eventlog also had them. They might also be useful in the instrumented runtime if we add some more high frequency events?